In [1]:
from sqlalchemy import create_engine, text
import pandas as pd 
import datetime as dt
import seaborn as sns
from matplotlib import pyplot as pyplot

In [2]:
database_name = "scooter"
connection_string = f"postgresql://postgres:postgres@localhost:5432/{database_name}"

In [3]:
engine = create_engine(connection_string)

In [4]:
bird = pd.read_csv('../data/scooters_bird.csv')
bolt = pd.read_csv('../data/scooters_bolt.csv')
gotcha = pd.read_csv('../data/scooters_gotcha.csv')
jump = pd.read_csv('../data/scooters_jump.csv')
lime = pd.read_csv('../data/scooters_lime.csv')
lyft = pd.read_csv('../data/scooters_lyft.csv')
spin = pd.read_csv('../data/scooters_spin.csv')
trips = pd.read_csv('../data/trips.csv')

In [5]:
pd.to_datetime(trips['startdate'].astype(str) + ' ' + trips['starttime'].astype(str), format = 'mixed')

0        2019-05-01 00:00:20.460000
1        2019-05-01 00:01:50.090000
2        2019-05-01 00:03:47.363333
3        2019-05-01 00:04:21.386666
4        2019-05-01 00:04:27.796666
                    ...            
565517   2019-07-31 23:47:43.000000
565518   2019-07-31 23:57:19.000000
565519   2019-07-31 23:41:52.000000
565520   2019-07-31 23:26:15.000000
565521   2019-07-31 22:19:08.000000
Length: 565522, dtype: datetime64[ns]

In [6]:
#converting columns to datetime & timedelta so they can be combined
trips["startdate"] = pd.to_datetime(trips["startdate"])
trips["starttime"] = pd.to_timedelta(trips["starttime"])

EQ1: Are there any null values in any columns in either table?

In [7]:
# Use "company.info(verbose=True, show_counts=True)" to investicage shapes of parsed data

print("---BIRD INFO---")
print(bird.info(verbose=True, show_counts=True))
print("---BOLT INFO---")
print(bolt.info(verbose=True, show_counts=True))
print("---GOTCHA INFO---")
print(gotcha.info(verbose=True, show_counts=True))
print("---JUMP INFO---")
print(jump.info(verbose=True, show_counts=True))
print("---LYFT INFO---")
print(lyft.info(verbose=True, show_counts=True))
print("---SPIN INFO---")
print(spin.info(verbose=True, show_counts=True))

## charge level: Bolt, Spin

---BIRD INFO---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12251590 entries, 0 to 12251589
Data columns (total 10 columns):
 #   Column       Non-Null Count     Dtype  
---  ------       --------------     -----  
 0   Unnamed: 0   12251590 non-null  int64  
 1   pubdatetime  12251590 non-null  object 
 2   latitude     12251590 non-null  float64
 3   longitude    12251590 non-null  float64
 4   sumdid       12251590 non-null  object 
 5   sumdtype     12251590 non-null  object 
 6   chargelevel  12251590 non-null  float64
 7   sumdgroup    12251590 non-null  object 
 8   costpermin   12251590 non-null  float64
 9   companyname  12251590 non-null  object 
dtypes: float64(4), int64(1), object(5)
memory usage: 934.7+ MB
None
---BOLT INFO---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3477198 entries, 0 to 3477197
Data columns (total 10 columns):
 #   Column       Non-Null Count    Dtype  
---  ------       --------------    -----  
 0   Unnamed: 0   3477198 non-null  int64  

EA1. Bolt and Spin have missing some missing values in the chargelevel column

EQ2. What date range is represented in each of the date columns? Investigate any values that seem odd.

In [8]:
# What date range is represented in each of the date columns? Investigate any values that seem odd.

#scooters - pubdatetime
scooters_date_query = '''SELECT pubdatetime FROM scooters'''
#trips - pubtimestamp, startdate, starttime, enddate, endtime, create_dt
trips_date_query = '''SELECT pubtimestamp, startdate, starttime, enddate, endtime, create_dt FROM trips'''

with engine.connect() as connection:
    scoot_dates = pd.read_sql(text(scooters_date_query), con = connection)
    trip_dates = pd.read_sql(text(trips_date_query), con = connection)

In [9]:
#scooter dates
#earliest
scootfirstdate = scoot_dates.min()
#latest
scootlastdate = scoot_dates.max()
#range
scootrange = scootlastdate - scootfirstdate

#trips dates
#earliest
trippubfirst = trip_dates.pubtimestamp.min()
tripstartfirst = trip_dates.startdate.min()
tripendfirst = trip_dates.enddate.min()
tripcratefirst = trip_dates.create_dt.min()
#latest dates
trippublast = trip_dates.pubtimestamp.max()
tripstartlast  = trip_dates.startdate.max()
tripendlast  = trip_dates.enddate.max()
tripcratelast  = trip_dates.create_dt.max()

#date ranges
trippubrange = trippublast - trippubfirst
tripstartrange  = tripstartlast - tripstartfirst
tripendrange  = tripendlast - tripendfirst
tripcraterange  = tripcratelast - tripcratefirst

In [10]:
print('scooter table date ranges')
print('scooter - pubdatetime - earliest: ' + str(scootfirstdate))
#latest
print('scooter - pubdatetime - latest: ' + str(scootlastdate))
#range
print('scooter - pubdatetime - range: ' + str(scootrange))

print('trips table date ranges')
#earliest
print('trips - pubdatetime - earliest: ' + str(trippubfirst))
print('trips - starttime - earliest: ' + str(tripstartfirst))
print('trips - endtime - earliest: ' + str(tripendfirst))
print('trips - create time - earliest: ' + str(tripcratefirst))

#latest dates
print('trips - pubdatetime - latest: ' + str(trippublast))
print('trips - starttime - latest: ' + str(tripstartlast))
print('trips - endtime - latest: ' + str(tripendlast))
print('trips - create time - latest: ' + str(tripcratelast))

#date ranges
print('trips - pubdatetime - range: ' + str(trippubrange))
print('trips - starttime - range: ' + str(tripstartrange))
print('trips - endtime - range: ' + str(tripendrange))
print('trips - create time - range: ' + str(tripcraterange))

scooter table date ranges
scooter - pubdatetime - earliest: pubdatetime   2019-05-01 00:01:41.247
dtype: datetime64[ns]
scooter - pubdatetime - latest: pubdatetime   2019-07-31 23:59:57
dtype: datetime64[ns]
scooter - pubdatetime - range: pubdatetime   91 days 23:58:15.753000
dtype: timedelta64[ns]
trips table date ranges
trips - pubdatetime - earliest: 2019-05-01 00:00:55.423000
trips - starttime - earliest: 2019-05-01
trips - endtime - earliest: 2019-05-01
trips - create time - earliest: 2019-05-02 05:30:23.780000
trips - pubdatetime - latest: 2019-08-01 07:04:00
trips - starttime - latest: 2019-07-31
trips - endtime - latest: 2019-08-01
trips - create time - latest: 2019-08-02 11:30:29.923000
trips - pubdatetime - range: 92 days 07:03:04.577000
trips - starttime - range: 91 days, 0:00:00
trips - endtime - range: 92 days, 0:00:00
trips - create time - range: 92 days 06:00:06.143000


Some fields are date only and format varies. Date range is from May 1, 2019 to July 31, 2019.

EQ3. Is time represented with am/pm or using 24 hour values in each of the columns that include time?

In [11]:
#investigate max dates
scoot_dates.pubdatetime.max()

Timestamp('2019-07-31 23:59:57')

In [12]:
trip_dates.starttime.max()

datetime.time(23, 59, 59, 506666)

Time is represented in 24 hour values: I%:M%:S%:f%

EQ4. What values are there in the sumdgroup column? Are there any that are not of interest for this project?

In [13]:
# What values are there in the sumdgroup column? Are there any that are not of interest for this project?
sumd_query = '''SELECT DISTINCT sumdgroup FROM scooters'''
with engine.connect() as connection:
    sumd_result = pd.read_sql(text(sumd_query), con = connection)
sumd_result

sumdgroup
0   bicycle
1   scooter
2   Scooter

In [14]:
# find bicycles 

bicycle_query = '''SELECT DISTINCT sumdid, companyname FROM scooters WHERE sumdgroup = 'bicycle';'''
with engine.connect() as connection:
    bikes = pd.read_sql(text(bicycle_query), con = connection)
bikes

sumdid companyname
0  Standard2UGJKREVB53HT        Lime
1  Standard5JXOV277MCWID        Lime
2  StandardNPOOZNUSGAXZN        Lime
3  StandardNUTLLXP4G37OI        Lime
4  StandardNW5HJFO4R32LY        Lime
5  StandardZPUQESHVPP74J        Lime

In [15]:
print(bikes.count())

sumdid         6
companyname    6
dtype: int64


EA4. The dataset contiains entires for six Lime bicycles; these entries should be omitted from scooter analysis

EQ5. What are the minimum and maximum values for all the latitude and longitude columns? Do these ranges make sense, or is there anything surprising?

In [16]:
# What are the minimum and maximum values for all the latitude and longitude columns? Do these ranges make sense, or is there anything surprising? -What is the range of values for trip duration and trip distance? Do these values make sense? Explore values that might seem questionable.

sll_query = '''SELECT
                MIN(latitude) as min_latitude,
                MAX(latitude) as max_latitude,
                MIN(longitude) as min_longitude,
                MAX(latitude) as max_longitude
            FROM scooters'''
            
tll_query = '''SELECT
                MIN(startlatitude) as min_latitude_start,
                MAX(startlatitude) as max_latitude_start,
                MIN(startlongitude) as min_longitude_start,
                MAX(startlongitude) as max_longitude_start,
                MIN(endlatitude) as min_latitude_end,
                MAX(endlatitude) as max_latitude_end,
                MIN(endlongitude) as min_longitude_end,
                MAX(endlongitude) as max_longitude_end
            FROM trips'''

with engine.connect() as connection:
    scoot_result = pd.read_sql(text(sll_query), con = connection)
    trip_result = pd.read_sql(text(tll_query), con = connection)

In [17]:
scoot_result

min_latitude  max_latitude  min_longitude  max_longitude
0           0.0  3.609874e+06     -97.443879   3.609874e+06

In [18]:
trip_result

min_latitude_start  max_latitude_start  min_longitude_start  \
0             35.8532           36.300029           -86.918008   

   max_longitude_start  min_latitude_end  max_latitude_end  min_longitude_end  \
0             -86.3662        -36.850405         51.045409        -122.673729   

   max_longitude_end  
0         174.764886

EA5. The scooters table has a minimum latitude value of 0.0. Latitude and longitude are expressed in defferent formats in the scooters and trips tables.

EQ6. Check out how the values for the company name column in the scooters table compare to those of the trips table. What do you notice?

In [19]:
# Check out how the values for the company name column in the scooters table compare to those of the trips table. What do you notice?
scooter_companies = '''SELECT DISTINCT companyname FROM scooters'''
trips_companies = ''' SELECT DISTINCT companyname FROM trips'''

with engine.connect() as connection:
    scompanies = pd.read_sql(text(scooter_companies), con = connection)
    tcompanies = pd.read_sql(text(trips_companies), con = connection)

In [20]:
scompanies.companyname.tolist()

['Bird', 'Bolt', 'Gotcha', 'Jump', 'Lime', 'Lyft', 'Spin']

In [21]:
tcompanies.companyname.tolist()

['Bird', 'Bolt Mobility', 'Gotcha', 'JUMP', 'Lime', 'Lyft', 'SPIN']

EA6. The same companies are present in the scooters and trips tables, however there are some variations in capitalization. Scooters contains company name "Bolt"; trips contains company name "Bolt Mobility"

Scooters: ['Bird', 'Bolt', 'Gotcha', 'Jump', 'Lime', 'Lyft', 'Spin']

Trips: ['Bird', 'Bolt Mobility', 'Gotcha', 'JUMP', 'Lime', 'Lyft', 'SPIN']

# During this period, seven companies offered scooters. How many scooters did each company have in this time frame? Did the number for each company change over time? Did scooter usage vary by company?

In [22]:
query = '''
SELECT companyname, count(DISTINCT(sumdid))
FROM scooters
WHERE sumdgroup NOT LIKE 'bicycle'
GROUP BY companyname
'''

with engine.connect() as connection:
    cc_distinctsumdid = pd.read_sql(text(query), con = connection)

In [23]:
cc_distinctsumdid.sort_values('count', ascending=False)

companyname  count
0        Bird   3860
4        Lime   1818
5        Lyft   1735
3        Jump   1210
6        Spin    805
1        Bolt    360
2      Gotcha    224

In [24]:
#Did the number for each company change over time?
query = '''SELECT companyname, COUNT(DISTINCT(sumdid)), CAST(pubdatetime as DATE) as date
            FROM scooters
            WHERE sumdgroup NOT LIKE 'bicycle'
            GROUP BY companyname, date;'''
with engine.connect() as connection:
    scooters_over_time = pd.read_sql(text(query), con = connection)
    
# scooters_over_time.date = pd.to_datetime(scooters_over_time.date, format = 'Y%-%m-%d')
    
scooters_over_time.sort_values('date', ascending=True)

companyname  count        date
0          Bird   1545  2019-05-01
316        Lime   1116  2019-05-01
158      Gotcha    148  2019-05-01
408        Lyft    674  2019-05-01
317        Lime   1149  2019-05-02
..          ...    ...         ...
88         Bird   1106  2019-07-31
315        Jump   1205  2019-07-31
499        Lyft    559  2019-07-31
407        Lime    500  2019-07-31
577        Spin    271  2019-07-31

[578 rows x 3 columns]

In [25]:
scooters_over_time.sort_values('date', ascending=True).reset_index()

index companyname  count        date
0        0        Bird   1545  2019-05-01
1      316        Lime   1116  2019-05-01
2      158      Gotcha    148  2019-05-01
3      408        Lyft    674  2019-05-01
4      317        Lime   1149  2019-05-02
..     ...         ...    ...         ...
573     88        Bird   1106  2019-07-31
574    315        Jump   1205  2019-07-31
575    499        Lyft    559  2019-07-31
576    407        Lime    500  2019-07-31
577    577        Spin    271  2019-07-31

[578 rows x 4 columns]

# According to Second Substitute Bill BL2018-1202 (as amended) (https://web.archive.org/web/20181019234657/https://www.nashville.gov/Metro-Clerk/Legislative/Ordinances/Details/7d2cf076-b12c-4645-a118-b530577c5ee8/2015-2019/BL2018-1202.aspx), all permitted operators will first clean data before providing or reporting data to Metro. Data processing and cleaning shall include:
Removal of staff servicing and test trips
Removal of trips below one minute
Trip lengths are capped at 24 hours
Are the scooter companies in compliance with the second and third part of this rule?

In [26]:
trips.head()

pubtimestamp companyname triprecordnum         sumdid  \
0  2019-05-01 00:00:55.423        Bird       BRD2134   Powered9EAJL   
1  2019-05-01 00:03:33.147        Lyft          LFT5  Powered296631   
2  2019-05-01 00:05:55.570        Bird       BRD2168   Powered7S2UU   
3  2019-05-01 00:05:55.570        Bird       BRD2166   PoweredZIIVX   
4  2019-05-01 00:05:55.570        Bird       BRD2165   PoweredJ7MB3   

   tripduration  tripdistance  startdate              starttime     enddate  \
0        3.0000     958.00528 2019-05-01 0 days 00:00:20.460000  2019-05-01   
1        1.7156    1371.39112 2019-05-01 0 days 00:01:50.090000  2019-05-01   
2        3.0000    2296.58800 2019-05-01 0 days 00:03:47.363333  2019-05-01   
3        3.0000    1200.78744 2019-05-01 0 days 00:04:21.386666  2019-05-01   
4        2.0000     351.04988 2019-05-01 0 days 00:04:27.796666  2019-05-01   

           endtime  startlatitude  startlongitude  endlatitude  endlongitude  \
0  00:02:52.346666       36.15710       -86.80360     36.15660     -86.80670   
1  00:03:33.026666       36.15797       -86.77896     36.16054     -86.77689   
2  00:07:13.596666       36.15470       -86.78180     36.15650     -86.78680   
3  00:06:59.176666       36.14940       -86.77950     36.15310     -86.77960   
4  00:06:23.150000       36.17780       -86.78660     36.17740     -86.78760   

                                           triproute                create_dt  
0  [(36.157235, -86.803612), (36.157235, -86.8036...  2019-05-02 05:30:23.780  
1  [(36.15797, -86.77896), (36.15795, -86.77873),...  2019-05-02 07:20:32.757  
2  [(36.155068, -86.782124), (36.156597, -86.78675)]  2019-05-02 05:30:24.530  
3  [(36.149741, -86.779344), (36.149741, -86.7793...  2019-05-02 05:30:24.237  
4  [(36.177699, -86.786477), (36.177711, -86.7864...  2019-05-02 05:30:24.207

In [27]:
print(trips.tripduration.min())
print(trips.tripduration.max())

-19.3582666667
512619.0


In [28]:
sub1 = trips[trips.tripduration < 1]
sub1.shape

(9154, 16)

In [29]:
over24 = trips[trips.tripduration > 1440]
over24.shape

(6938, 16)

In [30]:
trips.dtypes

pubtimestamp               object
companyname                object
triprecordnum              object
sumdid                     object
tripduration              float64
tripdistance              float64
startdate          datetime64[ns]
starttime         timedelta64[ns]
enddate                    object
endtime                    object
startlatitude             float64
startlongitude            float64
endlatitude               float64
endlongitude              float64
triproute                  object
create_dt                  object
dtype: object

In [31]:
#transform data

trips.pubtimestamp = pd.to_datetime(trips.pubtimestamp)

#combine datetimes to new columns and convert to datetime objects

trips['tripstart'] = pd.to_datetime(trips['startdate'].astype(str) + ' ' + trips['starttime'].astype(str), format = 'mixed')
trips['tripend'] = pd.to_datetime(trips['enddate'].astype(str) + ' ' + trips['endtime'].astype(str), format = 'mixed')

# trips['tripstart'] = str(trips['startdate']) + ' ' + trips['starttime']
# trips['tripend'] = trips['enddate'] + ' ' + trips['endtime']

print(trips.dtypes)

DateParseError: Unknown datetime string format, unable to parse: 2019-05-01 0 days 00:00:20.460000, at position 0

In [32]:
trips.head()

pubtimestamp companyname triprecordnum         sumdid  \
0 2019-05-01 00:00:55.423        Bird       BRD2134   Powered9EAJL   
1 2019-05-01 00:03:33.147        Lyft          LFT5  Powered296631   
2 2019-05-01 00:05:55.570        Bird       BRD2168   Powered7S2UU   
3 2019-05-01 00:05:55.570        Bird       BRD2166   PoweredZIIVX   
4 2019-05-01 00:05:55.570        Bird       BRD2165   PoweredJ7MB3   

   tripduration  tripdistance  startdate              starttime     enddate  \
0        3.0000     958.00528 2019-05-01 0 days 00:00:20.460000  2019-05-01   
1        1.7156    1371.39112 2019-05-01 0 days 00:01:50.090000  2019-05-01   
2        3.0000    2296.58800 2019-05-01 0 days 00:03:47.363333  2019-05-01   
3        3.0000    1200.78744 2019-05-01 0 days 00:04:21.386666  2019-05-01   
4        2.0000     351.04988 2019-05-01 0 days 00:04:27.796666  2019-05-01   

           endtime  startlatitude  startlongitude  endlatitude  endlongitude  \
0  00:02:52.346666       36.15710       -86.80360     36.15660     -86.80670   
1  00:03:33.026666       36.15797       -86.77896     36.16054     -86.77689   
2  00:07:13.596666       36.15470       -86.78180     36.15650     -86.78680   
3  00:06:59.176666       36.14940       -86.77950     36.15310     -86.77960   
4  00:06:23.150000       36.17780       -86.78660     36.17740     -86.78760   

                                           triproute                create_dt  
0  [(36.157235, -86.803612), (36.157235, -86.8036...  2019-05-02 05:30:23.780  
1  [(36.15797, -86.77896), (36.15795, -86.77873),...  2019-05-02 07:20:32.757  
2  [(36.155068, -86.782124), (36.156597, -86.78675)]  2019-05-02 05:30:24.530  
3  [(36.149741, -86.779344), (36.149741, -86.7793...  2019-05-02 05:30:24.237  
4  [(36.177699, -86.786477), (36.177711, -86.7864...  2019-05-02 05:30:24.207

In [33]:
trips.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 565522 entries, 0 to 565521
Data columns (total 16 columns):
 #   Column          Non-Null Count   Dtype          
---  ------          --------------   -----          
 0   pubtimestamp    565522 non-null  datetime64[ns] 
 1   companyname     565522 non-null  object         
 2   triprecordnum   565522 non-null  object         
 3   sumdid          565522 non-null  object         
 4   tripduration    565522 non-null  float64        
 5   tripdistance    565522 non-null  float64        
 6   startdate       565522 non-null  datetime64[ns] 
 7   starttime       565522 non-null  timedelta64[ns]
 8   enddate         565522 non-null  object         
 9   endtime         565522 non-null  object         
 10  startlatitude   565522 non-null  float64        
 11  startlongitude  565522 non-null  float64        
 12  endlatitude     565522 non-null  float64        
 13  endlongitude    565522 non-null  float64        
 14  triproute       5655

In [34]:
trips.dtypes

pubtimestamp       datetime64[ns]
companyname                object
triprecordnum              object
sumdid                     object
tripduration              float64
tripdistance              float64
startdate          datetime64[ns]
starttime         timedelta64[ns]
enddate                    object
endtime                    object
startlatitude             float64
startlongitude            float64
endlatitude               float64
endlongitude              float64
triproute                  object
create_dt                  object
dtype: object

In [35]:
trips.head()

pubtimestamp companyname triprecordnum         sumdid  \
0 2019-05-01 00:00:55.423        Bird       BRD2134   Powered9EAJL   
1 2019-05-01 00:03:33.147        Lyft          LFT5  Powered296631   
2 2019-05-01 00:05:55.570        Bird       BRD2168   Powered7S2UU   
3 2019-05-01 00:05:55.570        Bird       BRD2166   PoweredZIIVX   
4 2019-05-01 00:05:55.570        Bird       BRD2165   PoweredJ7MB3   

   tripduration  tripdistance  startdate              starttime     enddate  \
0        3.0000     958.00528 2019-05-01 0 days 00:00:20.460000  2019-05-01   
1        1.7156    1371.39112 2019-05-01 0 days 00:01:50.090000  2019-05-01   
2        3.0000    2296.58800 2019-05-01 0 days 00:03:47.363333  2019-05-01   
3        3.0000    1200.78744 2019-05-01 0 days 00:04:21.386666  2019-05-01   
4        2.0000     351.04988 2019-05-01 0 days 00:04:27.796666  2019-05-01   

           endtime  startlatitude  startlongitude  endlatitude  endlongitude  \
0  00:02:52.346666       36.15710       -86.80360     36.15660     -86.80670   
1  00:03:33.026666       36.15797       -86.77896     36.16054     -86.77689   
2  00:07:13.596666       36.15470       -86.78180     36.15650     -86.78680   
3  00:06:59.176666       36.14940       -86.77950     36.15310     -86.77960   
4  00:06:23.150000       36.17780       -86.78660     36.17740     -86.78760   

                                           triproute                create_dt  
0  [(36.157235, -86.803612), (36.157235, -86.8036...  2019-05-02 05:30:23.780  
1  [(36.15797, -86.77896), (36.15795, -86.77873),...  2019-05-02 07:20:32.757  
2  [(36.155068, -86.782124), (36.156597, -86.78675)]  2019-05-02 05:30:24.530  
3  [(36.149741, -86.779344), (36.149741, -86.7793...  2019-05-02 05:30:24.237  
4  [(36.177699, -86.786477), (36.177711, -86.7864...  2019-05-02 05:30:24.207

In [36]:
#calculate trip durations
trips['duration_full'] = trips['tripend'] - trips['tripstart']
trips.dtypes
trips['trip_minutes'] = trips.duration_full / dt.timedelta(minutes=1)

KeyError: 'tripend'

In [ ]:
#define compliance variables

#Trips Less Than One Minute
trips_sub1min = trips[trips['trip_minutes'] < 1]
print(trips_sub1min.info())

#Trips Greater than 24 Hours
trips_over24h = trips[trips['trip_minutes'] > 1440]
print(trips_over24h.info())

#Noncompliant Trips
trips_noncompliant = trips_sub1min._append(trips_over24h, ignore_index=True)

#noncompliant companies
noncompliant_companies = trips_noncompliant['companyname'].unique()

In [37]:
# DURATION ANALYSIS BY COMPNAY SUB MIN
(trips_sub1min
 .groupby(['companyname'])
 ['triprecordnum']
 .count()
 .reset_index()
 )

NameError: name 'trips_sub1min' is not defined

In [38]:
# DURATION ANALYSIS OVER 24 H
(trips_over24h
 .groupby(['companyname'])
 ['triprecordnum']
 .count()
 .reset_index()
 )

NameError: name 'trips_over24h' is not defined

In [39]:
# DURATION ANALYSIS ALL NONCOMPLIANT
(trips_noncompliant
 .groupby(['companyname'])
 ['triprecordnum']
 .count()
  .reset_index()
 .sort_values('triprecordnum', ascending=False)
 )

NameError: name 'trips_noncompliant' is not defined

In [40]:
#totals

totaltrips = len(trips.index)
totalsub1min = len(trips_sub1min.index)
totalover24h = len(trips_over24h.index)
totalnoncompliant = totalsub1min + totalover24h
percentagenoncompliant = totalnoncompliant/totaltrips * 100

print('Total number of scooter trips reported: ' + str(totaltrips))
print('Total number of scooter trips with a duration less than one minute: ' + str(totalsub1min))
print('Total number of scooter trips with a duration over 24 hours: ' + str(totalover24h))

print('Total number of noncompliant scooter trips: ' + str(totalnoncompliant))
print('Percentage of noncompliant scooter trips: ' + str(percentagenoncompliant))

print('Companies with noncompliant scooter trip data: ' + str(noncompliant_companies))

NameError: name 'trips_sub1min' is not defined

Total number of scooter trips reported: 565522
Total number of scooter trips with a duration less than one minute: 12384
Total number of scooter trips with a duration over 24 hours: 61
Total number of noncompliant scooter trips: 12445
Percentage of noncompliant scooter trips: 2.2006217264757164
Companies with noncompliant scooter trip data: ['Bird' 'Lyft' 'Lime' 'Bolt Mobility' 'SPIN' 'JUMP']
A2:
Total number of scooter trips reported: 565522
Total number of scooter trips with a duration less than one minute: 12384
Total number of scooter trips with a duration over 24 hours: 61
Total number of noncompliant scooter trips: 12445
Percentage of noncompliant scooter trips: 2.2006217264757164
Companies with noncompliant scooter trip data: ['Bird' 'Lyft' 'Lime' 'Bolt Mobility' 'SPIN' 'JUMP']

In [41]:
#pre update check
trips.shape
#Update the trips dataframe to remove any trips under one minute
trips = trips.drop(trips[trips.trip_minutes < 1].index)
#post update check
trips.shape
#Update the trips dataframe to remove any trips over 1440 minutes (24 hours = 1440 minutes)
trips = trips.drop(trips[trips.trip_minutes > 1440].index)
#post update check
trips.shape

AttributeError: 'DataFrame' object has no attribute 'trip_minutes'

In [ ]:
query = '''
SELECT *
FROM scooters
LIMIT 100;
''' 
with engine.connect() as connection:
    scooters = pd.read_sql(text(query), con = connection)

In [ ]:
scooters.shape

In [ ]:
scooters_may.shape

In [ ]:
scooters_june.shape

In [ ]:
scooters_july.shape

In [ ]:
# query = '''
# SELECT *
# FROM trips
# LIMIT 100;
# '''  
# with engine.connect() as connection:
#     trips = pd.read_sql(text(query), con = connection)

In [ ]:
trips.head()

In [ ]:
scooters.isna().any()

In [ ]:
# query = '''
# SELECT *
# FROM scooters
# WHERE chargelevel IS NULL
# '''  
# with engine.connect() as connection:
#     nullValues = pd.read_sql(text(query), con = connection)

In [ ]:
# nullValues.head()

In [ ]:
# nullValues

In [ ]:
query = '''
SELECT
    MAX(pubdatetime) - MIN(pubdatetime) AS range
    FROM scooters;
'''  
with engine.connect() as connection:
    daterangescooters = pd.read_sql(text(query), con = connection)

In [ ]:
daterangescooters.head()

In [ ]:
query = '''
SELECT
    MAX(pubtimestamp) - MIN(pubtimestamp) AS range
    FROM trips;
'''  
with engine.connect() as connection:
    daterangetrips = pd.read_sql(text(query), con = connection)

In [ ]:
daterangetrips.head()

In [ ]:
trips.isna().any()

In [ ]:
scooters['pubdatetime'] = pd.to_datetime(scooters['pubdatetime']) 
date_range = str(scooters['pubdatetime'].dt.date.min()) + ' to ' +str(scooters['pubdatetime'].dt.date.max())
print(date_range)

In [ ]:
trips['pubtimestamp'] = pd.to_datetime(trips['pubtimestamp']) 
date_range = str(trips['pubtimestamp'].dt.date.min()) + ' to ' +str(trips['pubtimestamp'].dt.date.max())
print(date_range)

In [ ]:
min_lat = trips['startlatitude'].min()
max_lat = trips['startlatitude'].max()
min_lon = trips['endlongitude'].min()
max_lon = trips['endlongitude'].max()
print(min_lat)
print(max_lat)
print(min_lon)
print(max_lon)

In [ ]:
merged_jump_scooters = pd.merge(jump, trips, on=['companyname'], how='union')
merged_jump_scooters.head()

In [ ]:
jump['pubdatetime'] = pd.to_datetime(jump['pubdatetime'])#, format = '%Y/%m/%d %I:%M:%S.%f')

In [ ]:
jump.dtypes

In [ ]:
trips['pubtimestamp'] = pd.to_datetime(trips['pubtimestamp'])#, format = '%Y/%m/%d %I:%M:%S.%f')

In [ ]:
trips.dtypes

In [ ]:
(trips.assign(month = trips['pubtimestamp'].dt.month_name())
.month
.value_counts(sort=False)
)

In [ ]:
avg_trips = trips.groupby(['sumdid', 'triprecordnum']).mean()

In [ ]:
avg_trips

In [ ]:
(trips.assign(date = trips['pubtimestamp'].dt.date)
.groupby(['date'])
['triprecordnum']
.count()
.reset_index()
.head(31)
)

In [ ]:
trips.head()

In [ ]:
trips.pubtimestamp = pd.to_datetime(trips.pubtimestamp)
trips.startdate = pd.to_datetime(trips.startdate)
trips.starttime = pd.to_datetime(trips.starttime, format = '%H:%M:%S.%f')
trips.enddate = pd.to_datetime(trips.enddate)
trips.endtime = pd.to_datetime(trips.endtime, format = '%I:%M:%S.%f')

In [ ]:
trips.dtypes

In [ ]:
trips.head()

In [ ]:
trips.groupby('sumdid').size()

In [ ]:
trips["sumdid"].value_counts()

In [ ]:
trips_days = (trips.assign(day = trips['pubtimestamp'].dt.day_name()).day.value_counts(sort = False) )

trips_days

(trips.groupby(pd.Grouper(key = 'pubtimestamp',
freq = '1d',
origin = 'epoch'
)) ['sumdid'].count().reset_index().head(10) )

(trips.assign(day_name = trips['pubtimestamp'].dt.day_name()).groupby('day_name')['sumdid'].nunique())

trips[['companyname', 'sumdid', 'pubtimestamp']].groupby(['companyname'])['sumdid'].nunique().reset_index(name = 'count').sort_values(['count'])

# The goal of Metro Nashville is to have each scooter used a minimum of 3 times per day. Based on the data, what is the average number of trips per scooter per day? Make sure to consider the days that a scooter was available. How does this vary by company?

In [42]:
values = []
for x in trips['companyname']:
    if x == "Bolt Mobility":
        values.append("Bolt")
    elif x == "JUMP":
        values.append("Jump")
    elif x == "SPIN":
        values.append("Spin")
    else :
        values.append(x)

In [43]:
trips['companyname'] = values

# Correct methodoloy

In [44]:
query = '''
    SELECT DISTINCT sumdid, companyname
    FROM scooters;
'''

with engine.connect() as connection:
    unique_scooters = pd.read_sql(text(query), con = connection)

In [45]:
unique_scooters.head()

sumdid companyname
0  Powered-017d3133-f14a-2b83-ee4f-d777e7c5b619        Bolt
1  Powered-01a24436-0315-e1bb-7ce0-d081d05dff7d        Bolt
2  Powered-03be23ca-d43b-222f-be54-e44b5b4690df        Bolt
3  Powered-046201fb-6532-1f37-6334-3612fb1e61f7        Bolt
4  Powered-0479bb84-afbd-0426-f1c4-df628542a88c        Bolt

In [46]:
unique_scooters['sumdid'].nunique()

10018

In [47]:
unique_scooters.groupby('companyname')['sumdid'].nunique()

companyname
Bird      3860
Bolt       360
Gotcha     224
Jump      1210
Lime      1824
Lyft      1735
Spin       805
Name: sumdid, dtype: int64

In [48]:
query = '''
SELECT DISTINCT
    sumdid,
    companyname,
    EXTRACT(MONTH from pubdatetime) AS month,
    EXTRACT(DAY from pubdatetime) AS day
FROM scooters'''

with engine.connect() as connection:
    unique_scooters_by_date = pd.read_sql(text(query), con = connection)

In [49]:
unique_scooters_by_date.shape

(413563, 4)

# The goal of Metro Nashville is to have each scooter used a minimum of 3 times per day. Based on the data, what is the average number of trips per scooter per day? Make sure to consider the days that a scooter was available. How does this vary by company?

In [50]:
# query = '''
# SELECT
# 	DISTINCT sumdid AS Unit,
# 	startdate AS trip_date,
# 	COUNT(triprecordnum) AS trips,
# 	companyname
# FROM
# 	trips
# GROUP BY
# 	sumdid, trip_date, companyname
# ORDER BY
# 	trip_date;
# '''  
# with engine.connect() as connection:
#     trips_per_unit = pd.read_sql(text(query), con = connection)

In [51]:
# trips_per_unit[trips_per_unit['unit']=='Powered5129907']

In [52]:
query = '''
SELECT
    sumdid,
    COUNT(*) AS num_trips
    FROM trips
    GROUP BY sumdid;
'''

with engine.connect() as connection:
    trips_per_scooter = pd.read_sql(text(query), con = connection)

In [53]:
trips_per_scooter

sumdid  num_trips
0                                  Powered5129907         17
1                                    Powered5DQX9         92
2                                   Powered661244         72
3                                      Powered816          4
4     Powered0612f166-ea11-5de5-9563-50fd38cd2e7e          2
...                                           ...        ...
9000                                 PoweredKZZ55         63
9001                                 PoweredRT6M7         55
9002                                 PoweredUARJJ         25
9003                                 PoweredTSDBK         31
9004                         PoweredBP45IVHPZYERL        215

[9005 rows x 2 columns]

In [54]:
count_of_scooter_trips = trips_per_scooter.groupby('sumdid')['num_trips'].sum().reset_index()

In [55]:
count_of_scooter_trips

sumdid  num_trips
0     Powered-017d3133-f14a-2b83-ee4f-d777e7c5b619         92
1     Powered-01a24436-0315-e1bb-7ce0-d081d05dff7d         57
2     Powered-03be23ca-d43b-222f-be54-e44b5b4690df         81
3     Powered-046201fb-6532-1f37-6334-3612fb1e61f7         71
4     Powered-0479bb84-afbd-0426-f1c4-df628542a88c         96
...                                            ...        ...
9000   Poweredff356f94-2252-5432-b484-ad9679bf4b8d         11
9001   Poweredff4b01f3-1f3f-5d6b-b1c5-5d6628ffa629          6
9002   Poweredff8a28ab-0ffb-59ce-9b1e-2d1299a25634         23
9003   Poweredffcc1b11-af95-5e06-b002-82ffb342aae8          6
9004                         StandardLIRPOZ7X3ZVGO          1

[9005 rows x 2 columns]

In [56]:
df = sns.load_dataset('count_of_scooter_trips')
 
# Use the 'hue' argument to provide a factor variable
sns.lmplot( x="trips", y="companyname", data=count_of_scooter_trips, fit_reg=False, hue='companyname', legend=False)
 
# Move the legend to an empty part of the plot
plt.legend(loc='lower right')

plt.show()

ValueError: 'count_of_scooter_trips' is not one of the example datasets.

In [65]:
avg_trips = trips_per_scooter.groupby('sumdid')['num_trips'].mean()

In [66]:
# avg_trips.sort_values(ascending=False).reset_index()

In [67]:
# avg_trips.reset_index()

In [68]:
avg_trips.reset_index()[avg_trips.reset_index()['unit']=='Powered5129907']

KeyError: 'unit'

In [61]:
unique_scooters_by_date.head()

sumdid companyname  month   day
0          Powered4JDA9        Bird    5.0   5.0
1  PoweredTZG4LZCRMDM4N        Lime    5.0   1.0
2          PoweredEMPQM        Bird    5.0   3.0
3  PoweredYZ2UREMTDYLB6        Lime    5.0  14.0
4          Powered15DW8        Bird    5.0  14.0

In [62]:
total_trips_and_days_by_company = (
    unique_scooters_by_date.groupby(['companyname', 'sumdid'])
    ['day'].count().reset_index().rename(columns = {'day': 'num_days'}).merge(trips_per_scooter).groupby('companyname')
    [['num_days', 'num_trips']].sum()
)
total_trips_and_days_by_company

num_days  num_trips
companyname                     
Bird           134437     149029
Bolt            16467      21890
Gotcha          12163       3315
Jump            23421       4529
Lime            76582     225656
Lyft            56208     120959
Spin            34307      34450

In [70]:
total_trips_and_days_by_company['num_trips'] / total_trips_and_days_by_company['num_days']

companyname
Bird      1.108542
Bolt      1.329325
Gotcha    0.272548
Jump      0.193373
Lime      2.946593
Lyft      2.151989
Spin      1.004168
dtype: float64

In [64]:
totals = total_trips_and_days_by_company.sum()
totals['num_trips'] / totals['num_days']

1.583291146400441

In [71]:
trips.head()

pubtimestamp companyname triprecordnum         sumdid  \
0 2019-05-01 00:00:55.423        Bird       BRD2134   Powered9EAJL   
1 2019-05-01 00:03:33.147        Lyft          LFT5  Powered296631   
2 2019-05-01 00:05:55.570        Bird       BRD2168   Powered7S2UU   
3 2019-05-01 00:05:55.570        Bird       BRD2166   PoweredZIIVX   
4 2019-05-01 00:05:55.570        Bird       BRD2165   PoweredJ7MB3   

   tripduration  tripdistance  startdate              starttime     enddate  \
0        3.0000     958.00528 2019-05-01 0 days 00:00:20.460000  2019-05-01   
1        1.7156    1371.39112 2019-05-01 0 days 00:01:50.090000  2019-05-01   
2        3.0000    2296.58800 2019-05-01 0 days 00:03:47.363333  2019-05-01   
3        3.0000    1200.78744 2019-05-01 0 days 00:04:21.386666  2019-05-01   
4        2.0000     351.04988 2019-05-01 0 days 00:04:27.796666  2019-05-01   

           endtime  startlatitude  startlongitude  endlatitude  endlongitude  \
0  00:02:52.346666       36.15710       -86.80360     36.15660     -86.80670   
1  00:03:33.026666       36.15797       -86.77896     36.16054     -86.77689   
2  00:07:13.596666       36.15470       -86.78180     36.15650     -86.78680   
3  00:06:59.176666       36.14940       -86.77950     36.15310     -86.77960   
4  00:06:23.150000       36.17780       -86.78660     36.17740     -86.78760   

                                           triproute                create_dt  
0  [(36.157235, -86.803612), (36.157235, -86.8036...  2019-05-02 05:30:23.780  
1  [(36.15797, -86.77896), (36.15795, -86.77873),...  2019-05-02 07:20:32.757  
2  [(36.155068, -86.782124), (36.156597, -86.78675)]  2019-05-02 05:30:24.530  
3  [(36.149741, -86.779344), (36.149741, -86.7793...  2019-05-02 05:30:24.237  
4  [(36.177699, -86.786477), (36.177711, -86.7864...  2019-05-02 05:30:24.207